### PR TITLE
Add Consumer and Delivery documentation

### PR DESCRIPTION
**Pull request by gerritsangel**
This adds some documentation to the Consumer and Delivery crates.

There might be some errors because of my understanding of the AMQP protocol.

### CODE REVIEW

File: src/consumer.rs [hunk 1]
```diff
@@ -37,6 +37,45 @@ impl<
     }
 }
 
+/// Continuously consumes message from a Queue
+///
+/// A consumer represents the [basic.consume](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.consume) AMQP command.
```
Comment (Keruspe) [inline src/consumer.rs:42]: It doesn't represent the command itself, but rather the stream of messages created by the command

File: src/consumer.rs [hunk 2]
```diff
@@ -37,6 +37,45 @@ impl<
...
+/// [basic.get](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.get) command, which
+/// retrieves only a single message.
+///
+/// A consumer is obtained by calling [`Channel::basic_consume`] with the queue name.
+/// New messages from this consumer can be accessed by obtaining the iterator from the consumer.
+/// This iterator returns new messages in the form of a
+/// [`Delivery`] for as long as the consumer is subscribed to the queue.
+///
+/// It is important to acknowledge each message if acknowledgments are not disabled by calling
+/// [`Channel::basic_ack`] with the delivery tag of the message.
```
Comment (Keruspe) [inline src/consumer.rs:53]: ack is not the only option, there is also nack and reject

File: src/consumer.rs [hunk 3]
```diff
@@ -37,6 +37,45 @@ impl<
...
+///
+/// A consumer represents the [basic.consume](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.consume) AMQP command.
+/// It continuously receives messages from the queue, as opposed to the
+/// [basic.get](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.get) command, which
+/// retrieves only a single message.
+///
+/// A consumer is obtained by calling [`Channel::basic_consume`] with the queue name.
+/// New messages from this consumer can be accessed by obtaining the iterator from the consumer.
+/// This iterator returns new messages in the form of a
+/// [`Delivery`] for as long as the consumer is subscribed to the queue.
```
Comment (Keruspe) [inline src/consumer.rs:50]: Might be wok mentioning basic.cancel to "unsubscribe"

File: src/consumer.rs [hunk 4]
```diff
@@ -37,6 +37,45 @@ impl<
...
 
+/// Continuously consumes message from a Queue
+///
+/// A consumer represents the [basic.consume](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.consume) AMQP command.
+/// It continuously receives messages from the queue, as opposed to the
+/// [basic.get](https://www.rabbitmq.com/amqp-0-9-1-quickref.html#basic.get) command, which
+/// retrieves only a single message.
+///
+/// A consumer is obtained by calling [`Channel::basic_consume`] with the queue name.
+/// New messages from this consumer can be accessed by obtaining the iterator from the consumer.
```
Comment (Keruspe) [inline src/consumer.rs:48]: There is also the `set_delegate` method which automatically spawns the delegate on the executor for each message (so you basically get parallel handling of messages)

File: src/consumer.rs [hunk 5]
```diff
@@ -37,6 +37,45 @@ impl<
...
+///
+/// consumer
+///      .for_each(move |delivery| {
+///         let delivery = delivery.expect("error caught in in consumer");
+///         channel
+///             .basic_ack(delivery.delivery_tag, BasicAckOptions::default())
+///             .map(|_| ())
+///     })
+///     .await
+/// ```
```
Comment (Keruspe) [inline src/consumer.rs:75]: This is probably simpler than the for_each: https://github.com/sozu-proxy/lapin/commit/ae196beea759c2411fa03de6a4c7fbbce184c782

File: src/consumer.rs (hunk 1 above)
Comment (gerritsangel) [reply src/consumer.rs]: Thanks, updated the documentation.

File: src/consumer.rs (hunk 2 above)
Comment (gerritsangel) [reply src/consumer.rs]: Updated and linked to Delivery for further information.

File: src/consumer.rs (hunk 3 above)
Comment (gerritsangel) [reply src/consumer.rs:52]: While adding this, I stumbled across one thing:

At https://www.rabbitmq.com/amqp-0-9-1-reference.html#basic.consume.consumer-tag it says that a consumer tag is being generated if the field is empty.

I think the API could be improved if the `pub fn tag(&self) -> &ShortString` method is also on the `Consumer` struct itself. I think it's not obvious that the user needs to access the inner struct.
Or is there some reason that it is hidden on the ConsumerInner? Maybe because of the Mutex?

Somehow the rust doc doesn't even generate the documentation for ConsumerInner, so you can only find about the `tag` method when working through the source code.

**Comment by gerritsangel** [comment]
I've updated it again, can you take another look?

File: src/consumer.rs (hunk 3 above)
Comment (Keruspe) [reply src/consumer.rs:52]: Added Consumre::tag, thanks for noticing it. ConsumerInner shouldn't be public, it was made such as a workarounf when lapin-futures was separated but it's no longer necessary

File: src/consumer.rs [hunk 6]
```diff
@@ -37,6 +37,81 @@ impl<
...
+/// * If the flag [`BasicConsumeOptions::no_ack`] is set to `true` while obtaining the consumer from
+///   [`Channel::basic_consume`], the server implicitely acknowledges each message after it has been
+///   sent.
+/// * If the flag [`BasicConsumeOptions::no_ack`] is set to `false`, a message has to be explicitely
+///   acknowledged or rejected with [`Channel::basic_ack`],
+///   [`Channel::basic_reject`] or [`Channel::basic_nack`]. See the documentation at [`Delivery`]
+///   for further information.
+///
+/// Also see the RabbitMQ documentation concerning
+/// [Consumer Acknowledgements and Publisher Confirms](https://www.rabbitmq.com/confirms.html).
```
Comment (Keruspe) [inline src/consumer.rs:69]: This is not related, it's the other way around. Publisher confirms are when you use channel.confirm_select, then the client waits for the server to send you some acks. It's not related to consumers at all

**Comment by Keruspe** [comment]
Now, you get the Channel alongside the Delivery, which simplifies acking the messages using the same Channel that received them

File: src/consumer.rs (hunk 6 above)
Comment (gerritsangel) [reply src/consumer.rs:69]: I will link this document: https://www.rabbitmq.com/consumers.html#acknowledgement-modes

File: src/consumer.rs [hunk 7]
```diff
@@ -37,6 +37,81 @@ impl<
...
+/// ## Cancel subscription
+///
+/// To stop receiving messages, call [`Channel::basic_cancel`] with the consumer tag of this
+/// consumer.
+///
+///
+/// ## Example
+/// ```rust,no_run
+/// let mut consumer = channel
+///     .clone()
```
Comment (Keruspe) [inline src/consumer.rs:88]: no need for clone

File: src/consumer.rs [hunk 8]
```diff
@@ -37,6 +37,81 @@ impl<
...
+///     .basic_consume(
+///         "hello",
+///         "my_consumer",
+///         BasicConsumeOptions::default(),
+///         FieldTable::default(),
+///     )
+///     .await?;
+///
+/// while let Some(delivery) = consumer.next().await {
```
Comment (Keruspe) [inline src/consumer.rs:97]: It's Some((channel, delivery)) now

**Comment by Keruspe** [comment]
Thanks!
